Lax I/O functions and I/O functions for bytes objects

RemingtonRohel opened the pull request:
# Scope

Modifies the functions for reading DMAP data to accept more diverse inputs and dispatch to `darn-dmap` dependency accordingly.

Each `read_[type]` function accepts two inputs: a source (either a path or a `bytes` object), and an optional mode (one of `"strict"`, `"lax"`, or `"sniff"`). `strict` mode is the default, where any non-compliance with the DMAP format in the source will cause the function to raise an Error. `lax` mode can be used to suppress these errors; in the case of a corrupt file, it will return the compliant records from the start of the file, as well as the byte index where the corrupt records start. `sniff` mode can be used with a file (i.e. source is a path) to read only the first record from the file. This could be useful for getting the CPID of the data within, for example, without having to read the whole file. All modes work on both compressed (with bzip2) and non-compressed sources.

Each `write_[type]` function is used to turn the records (in dictionary form) back into the DMAP binary format. These functions take a source (a list of dictionaries), as well as an optional path to an output file. If no path is specified, the binary data is returned as a `bytes` object. Otherwise, the data is written to the output file.

## Read function usage
All functions for reading DMAP files are structured similarly. As a reminder, these functions are `read_iqdat`, `read_rawacf`, `read_fitacf`, `read_grid`, `read_map`, `read_snd`, and `read_dmap`. 
```python
def read_dmap(source: Union[str, bytes], mode: str = "strict") -> Union[list[dict], tuple[list[dict], Union[None, int]]]:
    """
    Reads in DMAP data from `source`.

    Parameters
    ----------
    source: Union[str, bytes]
        Where to read data from. If input is of type `str`, this is interpreted as the path to a file.
        If input is of type `bytes`, this is interpreted as the raw data itself.
    mode: str
        Mode in which to read the data, either "strict" (default), "lax", or "sniff". In "strict" mode, any corruption
        in the data will raise an error. In "lax" mode, all valid records will be returned in a tuple along with
        the byte index of `source` where the corruption starts. In "sniff" mode, `source` must be a path, and
        only the first record will be read.

    Returns
    -------
    If `mode` is `strict`, returns `list[dict]` which is the parsed records.
    If `mode` is `lax`, returns `(list[dict], None | int)`, where the first element is the records which were parsed,
    and the second is the byte index where `source` was no longer a valid record of type `fmt`.
    If `mode` is `sniff`, returns `dict`, which is the first record.
    """
```

## Write function usage
All functions for writing DMAP files are structured similarly. As a reminder, these functions are `write_iqdat`, `write_rawacf`, `write_fitacf`, `write_grid`, `write_map`, `write_snd`, and `write_dmap`. 
```python
def write_dmap(source: list[dict], outfile: Union[None, str] = None) -> Union[None, bytes]:
    """
    Writes DMAP data from `source` to either a `bytes` object or to `outfile`.

    Parameters
    ----------
    source: list[dict]
        list of DMAP records as dictionaries.
    outfile: Union[None, str]
        If `None`, returns the data as a `bytes` object. If this is a string, then this is interpreted as a path
        and data will be written to the filesystem.
    """
```

## Test

See [the tutorial on our documentation](https://github.com/SuperDARN/pyDARNio/blob/lax_io/docs/user/SDarn.md) and run some of the examples from there.

RemingtonRohel committed:
Bumped darn-dmap dependency to v0.2.0 and created a wrapper around its API.

* `read_[type]` functions now accept kwarg `mode` which defaults to "strict", and also could be "lax". This will dispatch to either `read_[type]` or `read_[type]_lax` of darn-dmap, respectively.
* Positional input to `read_[type]` function supports either `str` or `bytes` input, and will dispatch to `read_[type]_bytes[_lax]` of darn-dmap if `bytes` object is input and depending on the `mode` input.
* Similarly for writing, `write_[type]` functions no longer _require_ a path to outfile; if not specified, the data is returned as a bytes object via dispatch to darn-dmap's `write_[type]_bytes` functions.

RemingtonRohel committed:
Merge `develop` into `lax_io`.

RemingtonRohel committed:
Add file sniffing to dmap functionality.

RemingtonRohel committed:
Updated docs with file sniffing.

JWiker commented:
Since this changes import paths from `pydarnio.dmap` to `pydarnio.dmap_wrapper` is it worth adding a deprecation shim in the old location or documenting this as a breaking change?

JWiker commented on src/pydarnio/dmap_wrapper.py:
```diff
@@ -0,0 +1,362 @@
+def read_dispatcher(source: Union[str, bytes], fmt: str, mode: str) -> Union[list[dict], tuple[list[dict], Union[None, int]]]:
+    """
+    Reads in DMAP data from `source`.
+
```
Is it worth adding something like this in each read_* function docstring, or is it not really necessary?

`Note: For large files (>100MB), prefer file path input over bytes objects to avoid memory issues. Bytes input is intended for small payloads or streaming use cases.`

RemingtonRohel replied:
I'm not sure if that's really necessary, either way all the data is getting loaded into memory.

JWiker commented on src/pydarnio/dmap_wrapper.py:
```diff
@@ -0,0 +1,362 @@
+    if not isinstance(source, bytes) and not isinstance(source, str):
+        raise TypeError(f"invalid type for `source` {type(source)}: expected `str` or `bytes`")
+
+    fn_name = (
```
Maybe worth adding a comment explaining this logic a bit? E.g.

`Construct the darn-dmap function name dynamically based on parameters:`
  `Pattern: [sniff|read]_[fmt][_bytes][_lax]`
  `Examples: read_fitacf, read_fitacf_bytes, read_fitacf_lax, sniff_fitacf`

RemingtonRohel replied:
That's a good idea, much easier for the next person who comes along.

JWiker commented on src/pydarnio/dmap_wrapper.py:
```diff
@@ -0,0 +1,362 @@
+    elif isinstance(outfile, str):
+        getattr(dmap, f"write_{fmt}")(source, outfile)
+    else:
+        raise TypeError(f"invalid type for `outfile` {type(source)}: expected `str` or `None`")
```
Replace `type(source)` with `type(outfile)`?

RemingtonRohel replied:
Done

JWiker commented:
Given this new functionality, is now a good time to add unit tests? E.g. tests covering all three modes, compression handling, error cases, bytes I/O parity... Or a new PR could be opened to to address all unit testing?

JWiker commented:
Does the darn-dmap library handle bzip2 compression detection via magic bytes (BZh signature) rather than relying solely on file extensions?

RemingtonRohel commented:
> Since this changes import paths from `pydarnio.dmap` to `pydarnio.dmap_wrapper` is it worth adding a deprecation shim in the old location or documenting this as a breaking change?

I don't think that it's needed, this will be part of a new major release so I think documentation is the better option. Everything that is commonly used is accessible under `pydarnio`, e.g. you call `pydarnio.read_dmap`, not `pydarnio.dmap_wrapper.read_dmap`.

> Does the darn-dmap library handle bzip2 compression detection via magic bytes (BZh signature) rather than relying solely on file extensions?

No but that's a great idea, that'll allow it to handle bz2 compression on the byte streams too. I wasn't aware of the magic byte signature, thanks for that!


> Given this new functionality, is now a good time to add unit tests? E.g. tests covering all three modes, compression handling, error cases, bytes I/O parity... Or a new PR could be opened to to address all unit testing?

I'm not sure what's best here. I've got tests on the `darn-dmap` dependency, so in a way this would be double-testing. Not a bad thing, just maybe a touch redundant. I'll put some effort into testing either way, I just might add more to `darn-dmap` and link the tests to this PR in some way, if that sounds like a reasonable plan.

RemingtonRohel committed:
Changes per code review.

* Bump darn-dmap version to 0.4.0
* Fix an incorrect error message
* Clarify code snippet in read_dispatcher()

RemingtonRohel commented:
I've added a lot of extra tests to the [darn-dmap](https://github.com/SuperDARNCanada/dmap/tree/v0.4.0) dependency for DMAP I/O. You can see the codebase at the included link. The unit tests are included at the bottom of each `*.rs` file where applicable, and integration tests are in `tests/tests.rs`. The integration tests are effectively the only sort of tests that we even could potentially add to pyDARNio, but they are already covered succinctly by `darn-dmap`.

JWiker reviewed (approved):
Looking good to me

egthomas commented:
@RemingtonRohel sorry, I didn't realize the changes in this pull request were (at least partially) intended to address feedback from the Data Distribution WG with respect to file checking - thank you for putting this together!

My recollections from those discussions, as well as my general experience as a user, is that the "lax" mode should probably be the default behavior instead of the "strict" rules.  For example, this statement in the documentation:

```
However, in certain use cases, the non-corrupted data before the bad record may be of use.
```

is something I might respectfully disagree with - I can't think of any cases where non-corrupted data would *not* be of use.

RemingtonRohel committed:
Changed default read mode to "lax".

RemingtonRohel commented:
@egthomas thank you for the review! I accept your feedback, my use case is primarily as a data manager instead of a scientist, so defaulting to strict mode is my bias showing through. I've switched it over to default to "lax" mode. That does necessitate a change to the docs accordingly, since "lax" mode returns a tuple. I've updated those accordingly as well. If there is no other recommendations, I'll merge this and push a new release promptly.
